internal/servicelog/parser.go

```diff
@@ -0,0 +1,127 @@
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>
+
+package servicelog
+
+import (
+	"bufio"
+	"bytes"
+	"errors"
+	"io"
+	"time"
+)
+
+var (
+	errParseFields  = errors.New("line has too few fields")
+	errParseTime    = errors.New("invalid log timestamp")
+	errParseService = errors.New("invalid log service name")
+)
+
+// Entry is a parsed log entry.
+type Entry struct {
+	Time    time.Time
+	Service string
+	Message string
+}
+
+// Parser parses and iterates over logs from a Reader until EOF (or another
+// error occurs). Each log parser
+type Parser struct {
+	r     io.Reader
+	br    *bufio.Reader
+	entry Entry
+	err   error
+}
+
+// NewParser creates a Parser with the given buffer size.
+func NewParser(r io.Reader, size int) *Parser {
+	return &Parser{
+		r:  r,
+		br: bufio.NewReaderSize(r, size),
+	}
+}
+
+// Reset resets the internal buffer (and clears any error).
+func (p *Parser) Reset() {
+	p.br.Reset(p.r)
+	p.err = nil
+}
+
+// Next parses the next log from the reader and reports whether another log
+// is available (false is returned on EOF or other read error).
+func (p *Parser) Next() bool {
+	for p.err == nil {
+		line, err := p.br.ReadSlice('\n')
+		if err != nil && !errors.Is(err, io.EOF) && !errors.Is(err, bufio.ErrBufferFull) {
+			// Non EOF (and non-buffer full) error, stop now
+			p.err = err
+			break
+		}
+		if len(line) == 0 {
+			p.err = io.EOF
+			break
+		}
+		if errors.Is(err, io.EOF) {
+			// EOF reached, stop iterating after processing line
+			p.err = err
+		}
+		entry, err := Parse(line)
+		if err == nil {
+			// Normal log line
+			p.entry = entry
+			return true
+		}
+		if !p.entry.Time.IsZero() {
+			// Partial log line due to long line or "(... output truncated ...)",
+			// use timestamp and service from previous entry.
+			p.entry.Message = string(line)
+			return true
+		}
+	}
+	return false
+}
+
+// Entry returns the current log entry (should only be called after Next
+// returns true).
+func (p *Parser) Entry() Entry {
+	return p.entry
+}
+
+// Err returns the last error that occurred (EOF is not considered an error).
+func (p *Parser) Err() error {
+	if errors.Is(p.err, io.EOF) {
+		return nil
+	}
+	return p.err
+}
+
+// Parse parses a log entry of the form
+// "2021-05-20T15:39:12.345Z [service] log message".
+func Parse(line []byte) (Entry, error) {
+	fields := bytes.SplitN(line, []byte(" "), 3)
+	if len(fields) != 3 {
+		return Entry{}, errParseFields
+	}
+	// .999 allows any number of fractional seconds (including none at all)
+	timestamp, err := time.Parse("2006-01-02T15:04:05.999Z07:00", string(fields[0]))
```

Comment:
Unfortunately they're slightly different: this uses `.999` instead of `.000`, because that means allow any number of fractional seconds. I wanted to do that in case we decide to use microseconds (or just seconds), the parser will still work fine. `.000` means exactly 3 digits.

Reply:
Fair enough, maybe make this a constant and place it next to the other one.

Reply:
I've made this a constant now (however, I've left them in parser.go and formatter.go next to the relevant code). Per discussion, I also un-exported `TimeFormat` and `TimeFormatRegex` as those are implementation details, especially now that the formatter and parser are all in the same package.